### PR TITLE
assets: convert usage of go-bindata to Go embed

### PR DESCRIPTION
**Pull request by alimy**
### Describe the pull request

add go:embed support to instead go-bindata.

https://github.com/gogs/gogs/issues/6846

### Checklist

- [x] I agree to follow the [Code of Conduct](https://go.dev/conduct) by submitting this pull request.
- [x] I have read and acknowledge the [Contributing guide](https://github.com/gogs/gogs/blob/main/.github/contributing.md).
- [x] I have added test cases to cover the new code.

### CODE REVIEW

File: conf/conf.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2021 The Gogs Authors. All rights reserved.
```
Comment (unknwon) [inline conf/conf.go:1]:
```suggestion
// Copyright 2022 The Gogs Authors. All rights reserved.
```

File: conf/conf.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+func trimPath(name string) string {
+	if name == "conf/" || name == "conf" {
+		return "."
+	}
+	return strings.TrimPrefix(name, "conf/")
+}
+
+// MustAsset is like Asset but panics when Asset would return an error.
+// It simplifies safe initialization of global variables.
+func MustAsset(name string) []byte {
```
Comment (unknwon) [inline conf/conf.go:26]: Any chance we don't need this wrappers anymore? I think `embed.FS` can be easily [converted to `http.FileSystem` using `http.FS`](https://sourcegraph.com/search?q=context:global+http.FS%28+lang:Go+&patternType=literal).

File: public/public.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+package public
+
+import (
+	"embed"
+	"net/http"
+)
+
+//go:embed assets css img js less plugins
+var embedFS embed.FS
```
Comment (unknwon) [inline public/public.go:13]: I'm pretty sure there are better syntax than this, like `**/*.css`, `**/*.js`, could you double check?

File: templates/templates.go [hunk 4]
```diff
@@ -1,23 +1,25 @@
-// Copyright 2020 The Gogs Authors. All rights reserved.
+// Copyright 2021 The Gogs Authors. All rights reserved.
```
Comment (unknwon) [inline templates/templates.go:1]: We only put the year of file creation, so this change should be reverted

File: templates/templates.go [hunk 5]
```diff
@@ -1,23 +1,25 @@
...
 	"io"
+	"io/fs"
 	"io/ioutil"
 	"path"
 	"strings"
 
-	"gopkg.in/macaron.v1"
-
 	"gogs.io/gogs/internal/osutil"
+	"gopkg.in/macaron.v1"
```
Comment (unknwon) [inline templates/templates.go:18]: This change should be reverted

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the PR! Please see inline comments.

File: conf/conf.go (hunk 2 above)
Comment (alimy) [reply conf/conf.go:27]: this function is used in internal/conf package in  config module just to process configure logic, not need to public outside use `http.FileSystem`. Keep this methed just make the source code have a smaller change.

File: public/public.go (hunk 3 above)
Comment (alimy) [reply public/public.go]: 对前端不太熟， 所以就暴力的把里面文件全包含了。
fixed: `//go:embed assets/* css/* img/* js/* plugins/*`

File: conf/conf.go (hunk 2 above)
Comment (unknwon) [reply conf/conf.go:27]: OK makes sense!

How much changes does it take for making `internal/conf` being able to directly consume a `embed.FS`?

File: public/public.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2021 The Gogs Authors. All rights reserved.
```
Comment (unknwon) [inline public/public.go:1]:
```suggestion
// Copyright 2022 The Gogs Authors. All rights reserved.
```

File: conf/conf.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+package conf
+
+import (
+	"embed"
+	"io/fs"
+	"strings"
+)
+
+//go:embed **/*
+//go:embed app.ini
```
Comment (unknwon) [inline conf/conf.go:14]:
```suggestion
//go:embed **/*
```

I don't think the second line is needed, shouldn't `**/*` include everything?

File: templates/templates.go [hunk 8]
```diff
@@ -17,7 +19,8 @@ import (
 	"gogs.io/gogs/internal/osutil"
 )
 
-//go:generate go-bindata -nomemcopy -nometadata -ignore="\\.DS_Store" -pkg=templates -prefix=../../../templates -debug=false -o=templates_gen.go ../../../templates/...
+//go:embed *.tmpl **/*
```
Comment (unknwon) [inline templates/templates.go:22]:
```suggestion
//go:embed **/*
```

ditto

**Review by unknwon — CHANGES_REQUESTED** [review]

File: public/public.go (hunk 3 above)
Comment (alimy) [reply public/public.go]: 需要包括所有目录(所有层级的子目录)的`.`开头文件与`_`开头文件的话, 可以使用`all:dir`去匹配， 参考[cmd/go: add //go:embed all:pattern](https://github.com/golang/go/commit/36dbf7f7e63f3738795bb04593c3c011e987d1f3)与[embed all:pattern源码注释](https://github.com/golang/go/blame/master/src/embed/embed.go#L85), 但是这是>=go1.18的功能。
最暴力的简单匹配:  `//go:embed all:assets all:css all:img all:js all:plugins` 
目前的匹配还是不能通过DeepSource的检测， 大佬有什么更好的匹配规则？请直接优化！

File: templates/templates.go (hunk 8 above)
Comment (alimy) [reply templates/templates.go:22]: just `//go:embed **/* ` will miss home.tmpl and install.tmpl file, so need add `*.tmpl` to match .tmpl file in current directory.

File: internal/conf/conf.go [hunk 9]
```diff
@@ -65,7 +89,7 @@ func Init(customConf string) error {
 	var err error
 	File, err = ini.LoadSources(ini.LoadOptions{
 		IgnoreInlineComment: true,
-	}, conf.MustAsset("conf/app.ini"))
+	}, MustAsset("conf/app.ini"))
```
Comment (unknwon) [inline internal/conf/conf.go:92]: To my understanding, all we need is just a call to `conf.EmbedFS.ReadFile("hello.txt")` here.

File: conf/conf.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+// license that can be found in the LICENSE file.
+
+package conf
+
+import (
+	"embed"
+)
+
+//go:embed app.ini **/*
+var EmbedFS embed.FS
```
Comment (unknwon) [inline conf/conf.go:12]:
```suggestion
var Files embed.FS
```

I feel like `EmbedFS` isn't a great name, `conf.Files` looks much nicer, what do you think?

File: internal/conf/conf.go (hunk 9 above)
Comment (alimy) [reply internal/conf/conf.go:92]: > To my understanding, all we need is just a call to conf.EmbedFS.ReadFile("hello.txt") here.

just won't to change internal/conf  `conf.MustAsset()` function to keep the same logic that someone call `conf.MustAsset()`.

Comment (alimy) [reply internal/conf/conf.go:92]: > There are only [two places calling conf.MustAsset](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/gogs/gogs%24+conf.MustAsset&patternType=literal), and both of them I think can be solved by calling conf.EmbedFS.ReadFile("hello.txt").

yes, that is. but `onf.EmbedFS.ReadFile("hello.txt")` will return error so need process error cause, `conf.MustAsset() ` wrap this cause in same logic that just panic when error occurs. how about  keep this wrap to make it simple,

Comment (alimy) [reply internal/conf/conf.go:92]: 改conf模块的初衷是 保持`conf.MustAsset(), conf.Asset(), conf.AssetDir()` 三个函数的逻辑不变，也就保持了其他调用这3个函数的地方都可以保持不变。 这里就有一个问题， 传进来的路径参数都有`conf/` 这个前缀， 这是为了保持与自定义配置文件conf/* 相匹配已经之前代码就是这么处理的。 所以conf目前的改法就是所以变更到这3个函数就结束，因此多加了一个`trimPath(name string)` 函数去 删掉多余的conf前缀， 我看着也碍眼， 其实最好的做法是如果是要embed conf就不不要传'conf/'前缀进来， 所以要改调用这三个函数的地方，修改传入参数的逻辑。

Comment (unknwon) [reply internal/conf/conf.go:92]: > > There are only [two places calling conf.MustAsset](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/gogs/gogs%24+conf.MustAsset&patternType=literal), and both of them I think can be solved by calling conf.EmbedFS.ReadFile("hello.txt").
> 
> yes, that is. but `onf.EmbedFS.ReadFile("hello.txt")` will return error so need process error cause, `conf.MustAsset() ` wrap this cause in same logic that just panic when error occurs. how about keep this wrap to make it simple,

I think having an error check is just fine, no need to have this wrapper anymore.

Comment (alimy) [reply internal/conf/conf.go]: >删掉多余的conf前缀， 我看着也碍眼， 其实最好的做法是如果需要embed conf就不要传'conf/'前缀进来， 所以要改调用这三个函数的地方，修改传入参数的逻辑。

>> You'll already do the embed within the conf directory, there will be no conf prefix in my understanding of how Go embed works.

现在把`conf.MustAsset(), conf.Asset()` 以及`trimPath(name string)` 都优化掉了， 只保留了`conf.AssetDir()`这个辅助函数。本地测试暂时没发现什么问题。

**Review by unknwon — APPROVED** [review]
Looking good!

**Pull request merged**